runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -113,12 +124,35 @@ public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
 
   @Override
   public BoundedReader<T> createReader(PipelineOptions options) throws IOException {
-    return createReader(options, null);
+    return getOrCreateReader(options, null);
   }
 
-  public BoundedReader<T> createReader(PipelineOptions options, CheckpointMarkT checkpointMark)
-      throws IOException {
-    return new Reader(source.createReader(options, checkpointMark));
+  @SuppressWarnings("unchecked")
+  public BoundedReader<T> getOrCreateReader(
+      PipelineOptions options,
+      CheckpointMarkT checkpointMark) throws IOException {
+    try {
+      initReaderCache((long) readerCacheInterval);
+      return (BoundedReader<T>) readerCache.get(this, new ReaderLoader(options, checkpointMark));
+    } catch (ExecutionException e) {
+      throw new RuntimeException("Failed to get or create reader", e);
+    }
+  }
+
+  private static final Object READER_CACHE_LOCK = new Object();
+
+  private void initReaderCache(long readerCacheInterval) {
+    if (readerCache == null) {
```

Comment:
I wanted to avoid locking `MicrobatchSource`, though a case could be made that this block of code resolves quickly and synchronize the method as you suggest for shorter and cleaner code.